### PR TITLE
golang extension: implement cluster_specifier_plugin to choose cluster dynamically

### LINKED ISSUE #25372: golang extension: implement cluster_specifier_plugin to choose cluster dynamically
refer: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-routeaction
No need to support async goroutine.

Yep, it's not an HTTP filter.
Users could use Go code to choose a cluster by implementing such an interface. similar to the current HTTP filter.
```
type ClusterSpecifier interface {
	Choose(RequestHeaderMap) string
}
```

### PR DESCRIPTION
**Pull request by doujiang24**
Commit Message:
golang extension: implement cluster_specifier_plugin to choose cluster dynamically
fix #25372 

Additional Description:
1. we could choose Host inside a cluster by using dynamic metadata, see #25371 ; but sometimes, we need to select a different cluster, i.e. dynamic_forward_proxy cluster vs STATIC cluster.
2. yep, `cluster_header` could be a way to reach it. but, we have to remove the hacked header if it's not expected from upstream.

so, adding a new plugin could be a proper choice.

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
finally! the dep-review CI has shown itself useful!

```console
  contrib/golang/filters/http/test/test_data/echo/go.mod » golang.org/x/net@0.0.0-20190311183353-d8887717615a – golang.org/x/net/http2 Denial of Service vulnerability (high severity)
    ↪ https://github.com/advisories/GHSA-69cg-p879-7622
  contrib/golang/filters/http/test/test_data/echo/go.mod » golang.org/x/net@0.0.0-20190311183353-d8887717615a – Uncontrolled Resource Consumption (low severity)
    ↪ https://github.com/advisories/GHSA-vvpx-j8f3-3w6h
  contrib/golang/http/cluster_specifier/test/test_data/simple/go.mod » golang.org/x/net@0.0.0-20190311183353-d8887717615a – golang.org/x/net/http2 Denial of Service vulnerability (high severity)
    ↪ https://github.com/advisories/GHSA-69cg-p879-7622
  contrib/golang/http/cluster_specifier/test/test_data/simple/go.mod » golang.org/x/net@0.0.0-20190311183353-d8887717615a – Uncontrolled Resource Consumption (low severity)
    ↪ https://github.com/advisories/GHSA-vvpx-j8f3-3w6h
```

https://github.com/envoyproxy/envoy/actions/runs/4240541310/jobs/7369672618#step:3:16

File: api/contrib/envoy/extensions/http/cluster_specifier/golang/v3alpha/golang.proto
Comment (wangfakang) [inline api/contrib/envoy/extensions/http/cluster_specifier/golang/v3alpha/golang.proto:27]: We can separate it into a common type, which can be reused with the Go extension API. such as
```
message Library {
string id = 1 [(validate.rules).string = {min_len: 1}];
string path = 2 [(validate.rules).string = {min_len: 1}];
}
```

File: contrib/golang/common/go/registry/BUILD
Comment (wangfakang) [inline contrib/golang/common/go/registry/BUILD:14]: format

File: contrib/golang/filters/http/test/test_data/passthrough/go.mod
Comment (doujiang24) [reply contrib/golang/filters/http/test/test_data/passthrough/go.mod:5]: I prefer to keep the new changes since the old version doesn't work for now, we'd better make it work for now. And change it when the extension moves to standard.

Since the test cases may also be examples for users, making it work for now could be a better choice.
Users also have to change it when the extension moves to standard.

File: api/contrib/envoy/extensions/http/cluster_specifier/golang/v3alpha/golang.proto
Comment (doujiang24) [reply api/contrib/envoy/extensions/http/cluster_specifier/golang/v3alpha/golang.proto:27]: I prefer to keep it simple now since we could only reuse two fields, maybe we could turn to reuse it when we move dso into bootstrap.

File: contrib/golang/http/cluster_specifier/source/cgo.cc
Comment (wangfakang) [inline contrib/golang/http/cluster_specifier/source/cgo.cc:18]: add `#ifdef __cplusplus`

same elsewhere

File: contrib/golang/http/cluster_specifier/source/go/pkg/api/cluster.go
Comment (wangfakang) [inline contrib/golang/http/cluster_specifier/source/go/pkg/api/cluster.go:37]: Get(key string) (string, bool)

File: contrib/golang/http/cluster_specifier/source/golang_cluster_specifier.cc
Comment (wangfakang) [inline contrib/golang/http/cluster_specifier/source/golang_cluster_specifier.cc:69]: info -> debug

same elsewhere

File: contrib/golang/http/cluster_specifier/source/golang_cluster_specifier.cc
Comment (wangfakang) [inline contrib/golang/http/cluster_specifier/source/golang_cluster_specifier.cc:68]: new_len <= 0

File: contrib/golang/http/cluster_specifier/source/golang_cluster_specifier.cc
Comment (wangfakang) [inline contrib/golang/http/cluster_specifier/source/golang_cluster_specifier.cc:79]: This needs to be protected, otherwise it will be a memory security vulnerability. Or directly limit a maximum length.

**Review by wangfakang — COMMENTED** [review]
Some comments need to be addressed, others LGTM

File: contrib/golang/common/dso/dso.h
Comment (wangfakang) [inline contrib/golang/common/dso/dso.h:56]:
```suggestion
private:
```

File: contrib/golang/common/dso/dso.cc
Comment (wangfakang) [inline contrib/golang/common/dso/dso.cc:33]: shared_ptr is used, so i don't think it's a issue.

File: contrib/golang/common/dso/dso.h
Comment (wangfakang) [inline contrib/golang/common/dso/dso.h:22]:
```suggestion
private:
```

File: contrib/golang/common/dso/dso.h
Comment (wangfakang) [inline contrib/golang/common/dso/dso.h:24]: `handler_` It can only be read-only externally, and now becomes writable in the derived class. So it is safer to provide a function for export it. such as:
```
void* handler() {
    return handler_;
}

void setLoad(bool loaded) { loaded_ = loaded; }
```

File: contrib/golang/common/dso/dso.h
Comment (wangfakang) [inline contrib/golang/common/dso/dso.h:64]: remove unused type `DsoInstancePtr`.

File: contrib/golang/common/dso/dso.cc
Comment (doujiang24) [reply contrib/golang/common/dso/dso.cc]: yep, you are right, also added comments for it, thanks.

File: contrib/golang/common/dso/dso.h
Comment (doujiang24) [reply contrib/golang/common/dso/dso.h:24]: I think we keep the same safety for `HttpFilterDsoInstance`, that no one could update `handler_` through `HttpFilterDsoInstance`'s API.

For `DsoInstance`, it's their responsibility to take care of `handle_` those who are inheriting `DsoInstance`, as well as using the private member.

So, I prefer using `protected` here.

**Review by wangfakang — DISMISSED** [review]
LGTM and ping @wbpcode

**Comment by doujiang24** [comment]
friendly ping @wbpcode 

@mattklein123 Could you please take a look too? since this includes the new cluster_specifiy_plugin API.

**Comment by doujiang24** [comment]
resolved the conflicts and friend ping, thanks.

**Comment by doujiang24** [comment]
friendly ping @wbpcode 

I know it's some big, and maybe we could review it step by step, i.e. start from the API proto?

File: contrib/golang/common/dso/dso.h
Comment (wbpcode) [inline contrib/golang/common/dso/dso.h:42]: Prefer `HttpFilterDsoImpl`.

File: contrib/golang/common/dso/dso.h
Comment (wbpcode) [inline contrib/golang/common/dso/dso.h:75]: ditto.

File: contrib/golang/common/dso/dso.h
Comment (wbpcode) [inline contrib/golang/common/dso/dso.h:135]: could `absl::flat_hash_map` be used here?

File: contrib/golang/router/cluster_specifier/source/cgo.cc
Comment (wbpcode) [inline contrib/golang/router/cluster_specifier/source/cgo.cc:28]: nit: it would be better if there are some more detailed comments to tell the reader the meaning of this func and parameters.

File: contrib/golang/router/cluster_specifier/source/go/pkg/api/capi.go
Comment (wbpcode) [inline contrib/golang/router/cluster_specifier/source/go/pkg/api/capi.go:23]: could these shared by the http filter and cluster_specifier?

File: contrib/golang/router/cluster_specifier/source/golang_cluster_specifier.cc
Comment (wbpcode) [inline contrib/golang/router/cluster_specifier/source/golang_cluster_specifier.cc:63]: could simple `while` loop be used here to replace this `goto`?

**Review by wbpcode — COMMENTED** [review]
Thanks for your great contributions. This is great and some minor comments are added to start.

Sorry I have no much background knowledge about the cgo and ffi. So, If I am wrong or misunderstand something, feel free to point it out.

**Comment by wbpcode** [comment]
And seems there is a refactor to the dso manager and dso interface. Could we split these changes to separated PR？

File: api/contrib/envoy/extensions/router/cluster_specifier/golang/v3alpha/golang.proto
Comment (doujiang24) [reply api/contrib/envoy/extensions/router/cluster_specifier/golang/v3alpha/golang.proto:50]: Not exactly, it's `plugin_config` in http filter,  I personally prefer to keep them separated, since they are both in the early stage, so that we could iterate them separately.

I'm fine if you think it's better to share the `library_id` and `library_path`, at this stage.

**Comment by doujiang24** [comment]
> And seems there is a refactor to the dso manager and dso interface. Could we split these changes to separated PR？

Split it to https://github.com/envoyproxy/envoy/pull/26116, please take a look there, thanks!

Thanks @wbpcode , all comments are addressed, please take another look, thanks!

**Comment by doujiang24** [comment]
friendly ping @wbpcode 

already resolved the merge conflicts, please take another look, thanks!

**Comment by alyssawilk** [comment]
@wbpcode ping?

**Comment by alyssawilk** [comment]
as a reminder @wbpcode AFIK maintainers aren't responsible for code quality in contrib, AFIK we just make sure CI is happy and contrib authors/reviewers are happy and you're good to merge.

**Comment by wbpcode** [comment]
got it. then I will just do a quick check tomorrow.

**Review by wbpcode — DISMISSED** [review]
LGTM. Thanks.

**Comment by wbpcode** [comment]
cc @doujiang24 could you resolve the conflicts first then we can merge it before the weekend? 😄

**Comment by doujiang24** [comment]
@wbpcode Thanks, resolved conflicts, glad to see this could be merged soon :)

**Comment by wbpcode** [comment]
> @wbpcode Thanks, resolved conflicts, glad to see this could be merged soon :)

Thanks for your create work.

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**
